compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            self.tcx.impl_trait_ref(obligation.impl_def_id);
+
+        let Some(impl_trait_self_ref) = impl_trait_self_ref else {
+            // It is possible that this is absent. In this case, we make no progress.
+            return Err(expr);
+        };
+
+        // We only really care about the `Self` type itself, which we extract from the ref.
+        let impl_self_ty: Ty<'tcx> = impl_trait_self_ref.self_ty();
+
+        let impl_predicates: ty::GenericPredicates<'tcx> =
+            self.tcx.predicates_of(obligation.impl_def_id);
+        let Some(impl_predicate_index) = obligation.impl_def_predicate_index else {
+            // We don't have the index, so we can only guess.
+            return Err(expr);
+        };
+
+        if impl_predicate_index >= impl_predicates.predicates.len() {
+            // This shouldn't happen, but since this is only a diagnostic improvement, avoid breaking things.
+            return Err(expr);
+        }
+        let relevant_broken_predicate: ty::PredicateKind<'tcx> =
+            impl_predicates.predicates[impl_predicate_index].0.kind().skip_binder();
+
+        // We want to find which of the generics in the `impl_generics` are relevant to
+        // the broken obligation predicate.
+        // A generic is relevant if it is mentioned in the "original" predicate. If we can't narrow it down, treat them all as relevant.
+        let relevant_impl_generics = impl_generics.params.iter();
+
+        let relevant_impl_generics: Vec<&ty::GenericParamDef> = match relevant_broken_predicate {
+            ty::PredicateKind::Clause(ty::Clause::Trait(broken_trait)) => relevant_impl_generics
+                .filter(|&generic| {
+                    // Only retain generics that are mentioned in the (Self type) of this predicate:
+                    find_param_def_in_ty_walker(broken_trait.trait_ref.self_ty().walk(), generic)
+                })
+                .collect(),
+            _ => {
+                // Because we are missing information about which might be relevant, we assume that they all are.
+                relevant_impl_generics.collect()
+            }
+        };
+
+        enum PointableType<'tcx> {
+            Adt(ty::AdtDef<'tcx>),
+            Tuple,
+        }
+
+        // Note that there is no point in supporting "primitive" types like char/i32,
+        // since we cannot refine such a span any more anyway.
+        // Right now, only ADTs (struct/enum (variant) constructors) and tuples are supported by this refinement.
+        let (impl_self_ty, impl_self_ty_args) = match impl_self_ty.kind() {
+            ty::Adt(impl_self_ty_path, impl_self_ty_args) => (
+                PointableType::Adt(*impl_self_ty_path),
+                impl_self_ty_args.iter().collect::<Vec<_>>(),
+            ),
+            ty::Tuple(impl_self_ty_args) => (
+                PointableType::Tuple,
+                impl_self_ty_args.iter().map(|t| ty::GenericArg::from(t)).collect::<Vec<_>>(),
```

Comment:
You can turn a list of ty into a list of generic args using [`as_substs`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.List.html#method.as_substs)